external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisCheckpointer.scala

```diff
@@ -64,7 +64,18 @@ private[kinesis] class KinesisCheckpointer(
   def removeCheckpointer(shardId: String, checkpointer: IRecordProcessorCheckpointer): Unit = {
     synchronized {
       checkpointers.remove(shardId)
-      checkpoint(shardId, checkpointer)
+    }
+    try {
+      // We must call `checkpoint()` with no parameter to finish reading shards.
+      // See an URL below for details:
+      // https://forums.aws.amazon.com/thread.jspa?threadID=244218
+      KinesisRecordProcessor.retryRandom(checkpointer.checkpoint(), 4, 100)
```

Comment:
Can you run this if `checkpointer != null`. The docs mention that the `checkpointer` may be null.